src/worker.rs

```diff
@@ -0,0 +1,320 @@
+use std::sync::Arc;
+use std::time::Duration;
+
+use chrono::{DateTime, Utc};
+use futures::future::{AbortHandle, Abortable, Aborted, BoxFuture, Future, FutureExt};
+use tokio::sync::oneshot;
+use tokio::task::{self, JoinHandle};
+use tokio::time;
+
+use crate::context::Context;
+
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum Restart {
+    Permanent,
+    Transient,
+    Temporary,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub enum Shutdown {
+    Indefinitely,
+    Timeout(Duration),
+}
+
+/// StartNotifier offers a convenient way to notify a worker spawner (a
+/// Supervisor in the general case) that the worker got started (or failed) to
+/// start.
+pub struct StartNotifier(Box<dyn FnOnce(Option<anyhow::Error>) + Send>);
```

Comment:
Couple things here. First, I think a `Result<(), Error>` is more idiomatic than `Option<Error>`. Second, using `anyhow::Error` is fine and probably much easier for initial development, but I think long term it would be better to use a generic parameter for the error, and trait bound it to `std::error::Error` like so:

```suggestion
pub struct StartNotifier<Error: std::error::Error>(Box<dyn FnOnce(Result<(), Error>) + Send>);
```

Reply:
What is the benefit of having an `Error` type parameter here, and not use a `Box<dyn>`? Mind you, all this is internal API and the building of this won't be exposed to the user (only success or failure).

Reply:
>  I think a Result<(), Error> is more idiomatic than Option<Error>

I'll update that type :+1:

Reply:
> What is the benefit of having an Error type parameter here, and not use a Box<dyn>? Mind you, all this is internal API and the building of values of this type won't be exposed to the API consumer (only success or failure).

That's a fair point, the suggestion of a type parameter was a bit reflexive. The more relevant point is that I think usage of `anyhow::Error` should be limited to application code. This is sort of an application-like usage though... I'm unsure. I think using `anyhow` for now is fine.

Reply:
Yep, I'm thinking we need to use `thiserror` for well known errors of this library, and keep API client errors in a box, could be anyhow or normal error. We should not force the `anyhow` API at the client API level (returned from client routines). That change should come later.